presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -536,6 +536,7 @@ protected void internalRefreshStats()
                     subGroup.internalRefreshStats();
                     cachedMemoryUsageBytes += subGroup.cachedMemoryUsageBytes;
                     if (!subGroup.isDirty()) {
+                        subGroup.updateEligiblility(); // final check if there's QUEUED queries
```

Comment:
I think it would be more correct to call `updateEligibility()` on line 542, after the if-else branch. This will only update when all the queries in the sub-group finish, which does fix the deadlock, but the resource group might have actually become eligible for scheduling earlier. Did you try that?